### PR TITLE
Add version of `!promotions nominate` command that does not require a rank to be specified

### LINKED ISSUE #162: Nominate command should not accept the target rank as a parameter
**Story**

As a user, I would like to nominate another user for the next rank in the line of progression without needing to specify what that rank is when nominating them.

### PR DESCRIPTION
**Pull request by Scott-Caldwell**
Fixes #162.

Adds a secondary version of the `!promotions nominate` command that does not require a rank to be specified.

This command has the form `!promotions nominate [user] [comment]`.

Upon typing the command, MODiX displays a confirmation message indicating the user and expected rank. MODiX directs the nominator to react with a :white_check_mark: in order to finalize the campaign creation. At the end of 10 seconds, if the user has reacted, then the campaign will be created. If the user has not reacted, then the emote will be replaced with a ❌ and the campaign will not be created.

I went with the above logic (using emotes) because I couldn't find a precedent in MODiX for anything similar, and it seemed like a simple, intuitive solution. Totally up for doing something different if we decide that there's a better way to go about it.

image
image

### CODE REVIEW

**Comment by mastorm** [comment]
I like the feel of it, but I am scared to merge. How far does this go up?
Can you theoretically promote up to admin?

Scott <notifications@github.com> schrieb am Sa. 10. Nov. 2018 um 21:54:

**Comment by Scott-Caldwell** [comment]
> I like the feel of it, but I am scared to merge. How far does this go up? Can you theoretically promote up to admin?

Sure. It still has the existing check for `Creating a promotion campaign requires a rank at least as high as the proposed target rank`, though, so it should still be fairly safe. I did try to make sure we kept the appropriate safeguards in place from the existing version of the command.

**Comment by mastorm** [comment]
We should some day add a configuration to it, but since you kept the safety
checks in tact, it’s not something this pr has to address. Conceptually
good to merge then, if the codereviews are done.

Scott <notifications@github.com> schrieb am Sa. 10. Nov. 2018 um 22:05:

File: Modix.Bot/Modules/PromotionsModule.cs [hunk 1]
```diff
@@ -67,6 +67,37 @@ public async Task Campaigns()
...
+            var reactionUsers = await confirmationMessage.GetReactionUsersAsync("✅");
+
+            if (reactionUsers.Any(u => u.Id == continuationData.NominatingUserId))
+            {
+                await PromotionsService.ContinueCreateCampaignAsync(continuationData);
+            }
+            else
+            {
+                await confirmationMessage.RemoveAllReactionsAsync();
+                await confirmationMessage.AddReactionAsync(new Emoji("❌"));
```
Comment (jrusbatch) [inline Modix.Bot/Modules/PromotionsModule.cs]: Just an idea - maybe it would be more clear to update the confirmation message's text to indicate that confirmation was not received in time and the campaign was not created.

**Comment by JakenVeina** [comment]
The promotion rules are baked into the service layer. It was by design that you can create promotions all the way up, but you have to A) have the `PromotionCampaignCreate` claim and B) be of the same rank or higher. Also C) only users with the `PromotionCampaignExecute` claim (I.E. staff) can actually execute promotions.

I very much like the UI concept here. I would tweak it to have MODiX automatically put both the checkmark and X emojis on the message, so it's more similar to a proper dialog. This is what we do with command errors today.

Also, we discussed at one point that the idea of having the target rank be a parameter in the service layer call actually makes no sense, since any attempt to create a campaign for a rank other than the next one up with throw an exception. I.E. there is only ever one rank that a user can be promoted to at any given time. It's useless to have it as a parameter. The only reason it IS a parameter is because the service call was written before that business rule was added.

Long story short, you should take the next step and actually remove the command that takes the targetRole parameter, and go remove that parameter from the service call.

File: Modix.Services/Promotions/PromotionsService.cs [hunk 2]
```diff
@@ -105,57 +124,78 @@ public class PromotionsService : IPromotionsService
         }
 
         /// <inheritdoc />
-        public async Task CreateCampaignAsync(ulong subjectId, ulong targetRoleId, string comment)
+        public async Task<CreateCampaignContinuationData> CreateCampaignAsync(ulong subjectId, string comment)
```
Comment (JakenVeina) [inline Modix.Services/Promotions/PromotionsService.cs]: Instead of having two methods for the two different "steps" of creating a campaign, I would just add a callback delegate as an optional parameter to this method, I.E. `Func<ProposedPromotionCampaign, Task<bool>> confirmDelegate`.

If you do this, make sure and change the `using` scope of the two transactions. I.E. don't `await` that callback inside the transaction scope, because it will ultimately prevent anyone from creating any campaigns, while a proposed campaign has yet to be confirmed or denied. The corollary is that if two users attempt to create a promotion at the same time, for the same user, one of them will get an exception. Oh well.

Also, yeah, in this context, I think an appropriate name for that new model class would be something like `ProposedPromotionCampaignBrief` or `ProposedPromotionCampaignSummary` or similar.

File: Modix.Services/Promotions/CreateCampaignContinuationData.cs [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+
+        public ulong NominatingUserId { get; }
+
+        public IEnumerable<GuildRoleBrief> RankRoles { get; }
+
+        public IGuildUser Subject { get; }
+
+        public GuildRoleBrief TargetRankRole { get; }
+
+        public void Deconstruct(
```
Comment (JakenVeina) [inline Modix.Services/Promotions/CreateCampaignContinuationData.cs]: Is this actually used anywhere? I certainly can't find it.

File: Modix.Services/Promotions/CreateCampaignContinuationData.cs [hunk 4]
```diff
@@ -0,0 +1,27 @@
+﻿using System.Collections.Generic;
+using Discord;
+using Modix.Data.Models.Core;
+
+namespace Modix.Services.Promotions
+{
+    public class CreateCampaignContinuationData
+    {
+        public CreateCampaignContinuationData(
```
Comment (JakenVeina) [inline Modix.Services/Promotions/CreateCampaignContinuationData.cs]: The current paradigm for model classes in MODiX is to just make them pure POCO's. `{ get; set; }` properties and default constructors. IMO, initializing an object through `{ }` initializer syntax is WAY cleaner than through a constructor, and access control modifiers exist for the sake of business logic, not data.

File: Modix.Bot/Modules/PromotionsModule.cs (hunk 1 above)
Comment (JakenVeina) [reply Modix.Bot/Modules/PromotionsModule.cs]: Definitely. The reactions make a lot more sense, I think, when we consider them to be a cheap knockoff for buttons. Which makes this message essentially a confirm dialog. We should add both reactions and see which one the user clicks. Then remove both of them and update/add a final result message

File: Modix.Services/Promotions/CreateCampaignContinuationData.cs (hunk 3 above)
Comment (jrusbatch) [reply Modix.Services/Promotions/CreateCampaignContinuationData.cs]: I did a double take, too, as this is the first time I've actually seen this used. This method is how [this line](https://github.com/discord-csharp/MODiX/pull/196/files/99b8fc7c16848a412c4713800783ae875a4f3f40#diff-9bd04b98f5322decb6467f3552ae18c0R11) works. See [Deconstructing user-defined types](https://docs.microsoft.com/en-us/dotnet/csharp/deconstruct#deconstructing-user-defined-types).

Comment (JakenVeina) [reply Modix.Services/Promotions/CreateCampaignContinuationData.cs]: Ahh, gotcha. I was only looking for it in the service methods.

Yeah, I'm definitely not a fan of that syntax, over just a traditional constructor that sets all the properties.

File: Modix.Bot/Modules/PromotionsModule.cs (hunk 1 above)
Comment (Scott-Caldwell) [reply Modix.Bot/Modules/PromotionsModule.cs]: Good idea. Implemented.

File: Modix.Services/Promotions/PromotionsService.cs (hunk 2 above)
Comment (Scott-Caldwell) [reply Modix.Services/Promotions/PromotionsService.cs]: I like it, makes the code feel less wonky. Updated.

File: Modix.Services/Promotions/CreateCampaignContinuationData.cs (hunk 3 above)
Comment (Scott-Caldwell) [reply Modix.Services/Promotions/CreateCampaignContinuationData.cs]: Haha, yeah, deconstructor is kinda weird. I made the class a POCO now, so no point in keeping the deconstructor anyway.

File: Modix.Services/Promotions/CreateCampaignContinuationData.cs (hunk 4 above)
Comment (Scott-Caldwell) [reply Modix.Services/Promotions/CreateCampaignContinuationData.cs]: Works for me! Updated.

File: Modix.Bot/Modules/PromotionsModule.cs [hunk 5]
```diff
@@ -69,15 +69,58 @@ public async Task Campaigns()
...
-            [Summary("The role for the user to be promoted to")]
-                IRole targetRole,
             [Remainder]
             [Summary("A comment to be attached to the new campaign")]
                 string comment)
-            => PromotionsService.CreateCampaignAsync(subject.Id, targetRole.Id, comment);
+        {
+            await PromotionsService.CreateCampaignAsync(subject.Id, comment, c => Confirm(c));
+
+            async Task<bool> Confirm(ProposedPromotionCampaignBrief proposedPromotionCampaign)
```
Comment (JakenVeina) [inline Modix.Bot/Modules/PromotionsModule.cs]: This is actually the first time I've ever seen a local function, out in the wild. Three cheers!
